File: backend/controllers/naruciteljiController.js
const naruciteljiService = require('../services/naruciteljiService');

class NaruciteljiController {
  async getAll(req, res) {
    try {
      const search = req.query.q || req.query.search || req.query.narucitelj || '';
      const hasGridParams = [
        'page',
        'pageSize',
        'sortBy',
        'sortOrder',
        'name',
        'mjesto',
        'drzava',
        'postanskiBroj',
        'OIB',
        'email'
      ].some((key) => req.query[key] !== undefined);

      const narucitelji = await naruciteljiService.getAllNarucitelji({
        search,
        page: req.query.page,
        pageSize: req.query.pageSize,
        sortBy: req.query.sortBy,
        sortOrder: req.query.sortOrder,
        filters: {
          name: req.query.name,
          mjesto: req.query.mjesto,
          drzava: req.query.drzava,
          postanskiBroj: req.query.postanskiBroj,
          OIB: req.query.OIB,
          email: req.query.email
        }
      });

      if (!hasGridParams) {
        res.json(narucitelji.items);
        return;
      }

      res.json(narucitelji);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  }

  async getById(req, res) {
    try {
      const narucitelj = await naruciteljiService.getNaruciteljiById(parseInt(req.params.id));
      res.json(narucitelj);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }

  async create(req, res) {
    try {
      const narucitelj = await naruciteljiService.createNarucitelj(req.body);
      res.status(201).json(narucitelj);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  async update(req, res) {
    try {
      const narucitelj = await naruciteljiService.updateNarucitelj(parseInt(req.params.id), req.body);
      res.json(narucitelj);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  async delete(req, res) {
    try {
      const narucitelj = await naruciteljiService.deleteNarucitelj(parseInt(req.params.id));
      res.json(narucitelj);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }
}

module.exports = new NaruciteljiController();

File: JavaSimpleBackend/src/main/java/com/atesti/controller/NaruciteljiController.java
package com.atesti.controller;

import com.atesti.dto.NaruciteljRequest;
import com.atesti.entity.Narucitelj;
import com.atesti.entity.RadniNalog;
import com.atesti.service.NaruciteljiService;
import com.atesti.service.RadniNaloziService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/narucitelji")
@RequiredArgsConstructor
public class NaruciteljiController {

    private final NaruciteljiService naruciteljiService;
    private final RadniNaloziService radniNaloziService;

    @GetMapping
    public ResponseEntity<?> getAll(
            @RequestParam(required = false) String q,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String narucitelj,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortOrder,
            @RequestParam(required = false) String name,
            @RequestParam(required = false) String mjesto,
            @RequestParam(required = false) String drzava,
            @RequestParam(required = false) String postanskiBroj,
            @RequestParam(required = false) String OIB,
            @RequestParam(required = false) String email) {

        String searchTerm = q != null ? q : (search != null ? search : (narucitelj != null ? narucitelj : ""));

        Object result = naruciteljiService.getAllNarucitelji(searchTerm, page, pageSize,
                sortBy, sortOrder, name, mjesto, drzava, postanskiBroj, OIB, email);

        return ResponseEntity.ok(result);
    }

    @GetMapping("/{id}")
    public ResponseEntity<Narucitelj> getById(@PathVariable Long id) {
        return ResponseEntity.ok(naruciteljiService.getNaruciteljiById(id));
    }

    @PostMapping
    public ResponseEntity<Narucitelj> create(@RequestBody NaruciteljRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(naruciteljiService.createNarucitelj(request));
    }

    @PutMapping("/{id}")
    public ResponseEntity<Narucitelj> update(@PathVariable Long id, @RequestBody NaruciteljRequest request) {
        return ResponseEntity.ok(naruciteljiService.updateNarucitelj(id, request));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Narucitelj> delete(@PathVariable Long id) {
        return ResponseEntity.ok(naruciteljiService.deleteNarucitelj(id));
    }

    @GetMapping("/{naruciteljiId}/radni-nalozi")
    public ResponseEntity<List<RadniNalog>> getRadniNaloziByNarucitelj(@PathVariable Long naruciteljiId) {
        return ResponseEntity.ok(radniNaloziService.getRadniNaloziByNarucitelj(naruciteljiId));
    }
}
